src/System.Management.Automation/engine/remoting/common/RunspaceConnectionInfo.cs

```diff
@@ -2440,48 +2440,58 @@ private static unsafe void FreeArray(byte** arr, int length)
-            stdOutReaderVar = null;
-            stdErrReaderVar = null;
-            try
-            {
-                stdInWriterVar = new StreamWriter(new NamedPipeServerStream(PipeDirection.Out, true, true, stdInPipeServer));
-                stdOutReaderVar = new StreamReader(new NamedPipeServerStream(PipeDirection.In, true, true, stdOutPipeServer));
-                stdErrReaderVar = new StreamReader(new NamedPipeServerStream(PipeDirection.In, true, true, stdErrPipeServer));
-            }
-            catch (Exception)
-            {
-                if (stdInWriterVar != null) { stdInWriterVar.Dispose(); } else { stdInPipeServer.Dispose(); }
+                try
+                {
+                    sshProcess = CreateProcessWithRedirectedStd(
+                        startInfo,
+                        out stdInPipeServer,
+                        out stdOutPipeServer,
+                        out stdErrPipeServer);
+                }
+                catch (InvalidOperationException e) { ex = e; }
+                catch (ArgumentException e) { ex = e; }
+                catch (FileNotFoundException e) { ex = e; }
+                catch (System.ComponentModel.Win32Exception e) { ex = e; }
+
+                if ((ex != null) ||
+                    (sshProcess == null) ||
+                    (sshProcess.HasExited))
+                {
+                    throw new InvalidOperationException(
+                        StringUtil.Format(RemotingErrorIdStrings.CannotStartSSHClient, (ex != null) ? ex.Message : string.Empty),
+                        ex);
+                }
 
-                if (stdOutReaderVar != null) { stdInWriterVar.Dispose(); } else { stdOutPipeServer.Dispose(); }
+                // Create the std in writer/readers needed for communication with ssh.exe.
+                stdInWriterVar = null;
+                stdOutReaderVar = null;
+                stdErrReaderVar = null;
+                try
+                {
+                    stdInWriterVar = new StreamWriter(new NamedPipeServerStream(PipeDirection.Out, true, true, stdInPipeServer));
+                    stdOutReaderVar = new StreamReader(new NamedPipeServerStream(PipeDirection.In, true, true, stdOutPipeServer));
+                    stdErrReaderVar = new StreamReader(new NamedPipeServerStream(PipeDirection.In, true, true, stdErrPipeServer));
+                }
+                finally
+                {
+                    stdInWriterVar?.Dispose();
+                    stdOutReaderVar?.Dispose();
+                    stdErrReaderVar?.Dispose();
+                }
 
-                if (stdErrReaderVar != null) { stdInWriterVar.Dispose(); } else { stdErrPipeServer.Dispose(); }
+                stdInPipeServer = null;
+                stdOutPipeServer = null;
+                stdErrPipeServer = null;
 
-                throw;
+                return sshProcess.Id;
+            }
+            finally
+            {
+                stdInPipeServer?.Dispose();
```

Comment:
These objects cannot be disposed here, as there are used in the active transport.  They are disposed here after the connection is closed.
https://github.com/PowerShell/PowerShell/blob/9aede0558a4c26a4f0769a51d8f0bb3e5fcfa744/src/System.Management.Automation/engine/remoting/fanin/OutOfProcTransportManager.cs#L1663

Reply:
Since we assigned to null above, these objects will not be disposed, unless an exception is thrown.

Reply:
Oh, yes I see that now.

Reply:
But these changes strike me as unnecessary, while making the code harder to read and maintain, just to mitigate a compiler warning that doesn't quite apply.  But I do agree that we should use the new null conditional syntax where applicable.
```csharp
stdInPipeServer?.Dispose();
```

Also I agree with @rjmholt about adding comments to make clear when the code disposes out objects (on error).